File: lld/problems/ride-sharing/go/ride_sharing_test.go
package ridesharing

import (
	"sync"
	"testing"
)

func newTestSystem() *RideSharingSystem {
	drivers := []*Driver{
		{ID: "D1", Name: "Alice", Location: Location{X: 0, Y: 0}, available: true},
		{ID: "D2", Name: "Bob", Location: Location{X: 10, Y: 10}, available: true},
	}
	return NewRideSharingSystem(drivers, NearestAvailableDriverStrategy{}, DistanceBasedPricing{BaseFare: 2, PerDistance: 1.5})
}

func TestHappyPathLifecycle(t *testing.T) {
	sys := newTestSystem()
	rider := &Rider{ID: "R1", Name: "Riya"}

	trip := sys.RequestRide(rider, Location{X: 0, Y: 0}, Location{X: 3, Y: 4})
	if trip.Status != Requested {
		t.Fatalf("expected Requested, got %v", trip.Status)
	}

	driver, err := sys.MatchDriver(trip.ID)
	if err != nil {
		t.Fatalf("expected match, got err: %v", err)
	}
	if driver.ID != "D1" {
		t.Fatalf("expected nearest driver D1, got %s", driver.ID)
	}
	if trip.Status != Accepted {
		t.Fatalf("expected Accepted, got %v", trip.Status)
	}

	if err := sys.StartTrip(trip.ID); err != nil {
		t.Fatalf("unexpected err starting trip: %v", err)
	}
	if trip.Status != InProgress {
		t.Fatalf("expected InProgress, got %v", trip.Status)
	}

	fare, err := sys.CompleteTrip(trip.ID)
	if err != nil {
		t.Fatalf("unexpected err completing trip: %v", err)
	}
	// base 2 + 5 (distance from (0,0) to (3,4)) * 1.5 = 9.5
	if fare != 9.5 {
		t.Fatalf("expected fare 9.5, got %v", fare)
	}
	if trip.Status != Completed {
		t.Fatalf("expected Completed, got %v", trip.Status)
	}
	if !driver.available {
		t.Fatalf("expected driver freed after trip completion")
	}
}

func TestInvalidTransitionRejected(t *testing.T) {
	sys := newTestSystem()
	rider := &Rider{ID: "R1", Name: "Riya"}
	trip := sys.RequestRide(rider, Location{X: 0, Y: 0}, Location{X: 1, Y: 1})

	if _, err := sys.CompleteTrip(trip.ID); err != ErrInvalidTransition {
		t.Fatalf("expected ErrInvalidTransition completing a Requested trip, got %v", err)
	}

	if _, err := sys.MatchDriver(trip.ID); err != nil {
		t.Fatalf("unexpected err matching driver: %v", err)
	}
	if err := sys.StartTrip(trip.ID); err != nil {
		t.Fatalf("unexpected err starting trip: %v", err)
	}
	if _, err := sys.CompleteTrip(trip.ID); err != nil {
		t.Fatalf("unexpected err completing trip: %v", err)
	}

	if err := sys.CancelTrip(trip.ID); err != ErrInvalidTransition {
		t.Fatalf("expected ErrInvalidTransition cancelling a Completed trip, got %v", err)
	}
}

func TestNoDriverAvailable(t *testing.T) {
	sys := newTestSystem()
	rider := &Rider{ID: "R1", Name: "Riya"}

	t1 := sys.RequestRide(rider, Location{X: 0, Y: 0}, Location{X: 1, Y: 1})
	t2 := sys.RequestRide(rider, Location{X: 0, Y: 0}, Location{X: 1, Y: 1})
	t3 := sys.RequestRide(rider, Location{X: 0, Y: 0}, Location{X: 1, Y: 1})

	if _, err := sys.MatchDriver(t1.ID); err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	if _, err := sys.MatchDriver(t2.ID); err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	if _, err := sys.MatchDriver(t3.ID); err != ErrNoDriverAvailable {
		t.Fatalf("expected ErrNoDriverAvailable, got %v", err)
	}
}

// TestConcurrentMatching asserts two goroutines racing for the same single
// available driver never both succeed — the mutex in MatchDriver must serialize them.
func TestConcurrentMatching(t *testing.T) {
	drivers := []*Driver{
		{ID: "D1", Name: "Alice", Location: Location{X: 0, Y: 0}, available: true},
	}
	sys := NewRideSharingSystem(drivers, NearestAvailableDriverStrategy{}, DistanceBasedPricing{BaseFare: 2, PerDistance: 1.5})
	rider := &Rider{ID: "R1", Name: "Riya"}

	trip1 := sys.RequestRide(rider, Location{X: 0, Y: 0}, Location{X: 1, Y: 1})
	trip2 := sys.RequestRide(rider, Location{X: 0, Y: 0}, Location{X: 1, Y: 1})

	var wg sync.WaitGroup
	results := make(chan error, 2)

	for _, tripID := range []string{trip1.ID, trip2.ID} {
		wg.Add(1)
		go func(id string) {
			defer wg.Done()
			_, err := sys.MatchDriver(id)
			results <- err
		}(tripID)
	}
	wg.Wait()
	close(results)

	successCount := 0
	for err := range results {
		if err == nil {
			successCount++
		}
	}
	if successCount != 1 {
		t.Fatalf("expected exactly 1 success for single available driver, got %d", successCount)
	}
}

File: lld/problems/ride-sharing/java/src/RideSharingTest.java
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Plain assert-based test runner (no JUnit dependency needed).
 * Run via `java -cp out Main` (invoked from Main) or `java -cp out RideSharingTest` directly.
 */
public class RideSharingTest {

    public static void main(String[] args) {
        runAll();
    }

    public static void runAll() {
        testHappyPathLifecycle();
        testInvalidTransitionRejected();
        testNoDriverAvailable();
        testConcurrentMatching();
        System.out.println("All RideSharingTest cases passed.");
    }

    private static RideSharingSystem newTestSystem() {
        List<Driver> drivers = new ArrayList<>();
        drivers.add(new Driver("D1", "Alice", new Location(0, 0), true));
        drivers.add(new Driver("D2", "Bob", new Location(10, 10), true));
        return new RideSharingSystem(drivers, new NearestAvailableDriverStrategy(),
                new DistanceBasedPricing(2, 1.5));
    }

    private static void testHappyPathLifecycle() {
        try {
            RideSharingSystem sys = newTestSystem();
            Rider rider = new Rider("R1", "Riya");

            Trip trip = sys.requestRide(rider, new Location(0, 0), new Location(3, 4));
            assertEquals(TripStatus.REQUESTED, trip.status, "expected REQUESTED");

            Driver driver = sys.matchDriver(trip.id);
            assertEquals("D1", driver.id, "expected nearest driver D1");
            assertEquals(TripStatus.ACCEPTED, trip.status, "expected ACCEPTED");

            sys.startTrip(trip.id);
            assertEquals(TripStatus.IN_PROGRESS, trip.status, "expected IN_PROGRESS");

            double fare = sys.completeTrip(trip.id);
            // base 2 + 5 (distance from (0,0) to (3,4)) * 1.5 = 9.5
            assertEquals(9.5, fare, "expected fare 9.5");
            assertEquals(TripStatus.COMPLETED, trip.status, "expected COMPLETED");
            assertTrue(driver.isAvailable(), "expected driver freed after trip completion");
        } catch (Exception e) {
            throw new AssertionError("unexpected exception: " + e, e);
        }
    }

    private static void testInvalidTransitionRejected() {
        RideSharingSystem sys = newTestSystem();
        Rider rider = new Rider("R1", "Riya");
        Trip trip = sys.requestRide(rider, new Location(0, 0), new Location(1, 1));

        try {
            sys.completeTrip(trip.id);
            throw new AssertionError("expected InvalidTransitionException completing a REQUESTED trip");
        } catch (InvalidTransitionException expected) {
            // expected
        } catch (Exception e) {
            throw new AssertionError("unexpected exception: " + e, e);
        }

        try {
            sys.matchDriver(trip.id);
            sys.startTrip(trip.id);
            sys.completeTrip(trip.id);
        } catch (Exception e) {
            throw new AssertionError("unexpected exception: " + e, e);
        }

        try {
            sys.cancelTrip(trip.id);
            throw new AssertionError("expected InvalidTransitionException cancelling a COMPLETED trip");
        } catch (InvalidTransitionException expected) {
            // expected
        } catch (Exception e) {
            throw new AssertionError("unexpected exception: " + e, e);
        }
    }

    private static void testNoDriverAvailable() {
        RideSharingSystem sys = newTestSystem();
        Rider rider = new Rider("R1", "Riya");

        Trip t1 = sys.requestRide(rider, new Location(0, 0), new Location(1, 1));
        Trip t2 = sys.requestRide(rider, new Location(0, 0), new Location(1, 1));
        Trip t3 = sys.requestRide(rider, new Location(0, 0), new Location(1, 1));

        try {
            sys.matchDriver(t1.id);
            sys.matchDriver(t2.id);
        } catch (Exception e) {
            throw new AssertionError("unexpected exception: " + e, e);
        }

        try {
            sys.matchDriver(t3.id);
            throw new AssertionError("expected NoDriverAvailableException");
        } catch (NoDriverAvailableException expected) {
            // expected
        } catch (Exception e) {
            throw new AssertionError("unexpected exception: " + e, e);
        }
    }

    // Asserts two threads racing for the same single available driver never
    // both succeed — the lock in matchDriver must serialize them.
    private static void testConcurrentMatching() {
        List<Driver> drivers = new ArrayList<>();
        drivers.add(new Driver("D1", "Alice", new Location(0, 0), true));
        RideSharingSystem sys = new RideSharingSystem(drivers, new NearestAvailableDriverStrategy(),
                new DistanceBasedPricing(2, 1.5));
        Rider rider = new Rider("R1", "Riya");

        Trip trip1 = sys.requestRide(rider, new Location(0, 0), new Location(1, 1));
        Trip trip2 = sys.requestRide(rider, new Location(0, 0), new Location(1, 1));

        String[] tripIds = new String[] { trip1.id, trip2.id };
        AtomicInteger successCount = new AtomicInteger(0);
        CountDownLatch latch = new CountDownLatch(tripIds.length);

        for (String tripId : tripIds) {
            Thread t = new Thread(() -> {
                try {
                    sys.matchDriver(tripId);
                    successCount.incrementAndGet();
                } catch (Exception ignored) {
                    // expected for the loser
                } finally {
                    latch.countDown();
                }
            });
            t.start();
        }

        try {
            latch.await();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }

        assertEquals(1, successCount.get(), "expected exactly 1 success for single available driver");
    }

    private static void assertTrue(boolean condition, String label) {
        if (!condition) {
            throw new AssertionError(label);
        }
    }

    private static void assertEquals(Object expected, Object actual, String label) {
        if (!expected.equals(actual)) {
            throw new AssertionError(label + ": expected " + expected + " but got " + actual);
        }
    }

    private static void assertEquals(double expected, double actual, String label) {
        if (Math.abs(expected - actual) > 1e-9) {
            throw new AssertionError(label + ": expected " + expected + " but got " + actual);
        }
    }
}
